File: mac/Sources/EdgeLinkKit/PairingWire.swift
import Foundation

public struct PairStartRequest: Codable, Sendable {
    public let hostId: String
    public let hostPk: String
    public let name: String

    public init(hostId: String, hostPk: String, name: String) {
        self.hostId = hostId
        self.hostPk = hostPk
        self.name = name
    }
}

public struct PairConfirmRequest: Codable, Sendable {
    public let role: String
    public let hostId: String
    public let clientId: String
    public let hostPk: String
    public let clientPk: String
    public let hostName: String
    public let clientName: String

    public init(role: String, hostId: String, clientId: String, hostPk: String, clientPk: String, hostName: String, clientName: String) {
        self.role = role
        self.hostId = hostId
        self.clientId = clientId
        self.hostPk = hostPk
        self.clientPk = clientPk
        self.hostName = hostName
        self.clientName = clientName
    }
}

public struct PairReadyBody: Codable, Sendable {
    public let deviceId: String
}

public struct PairCommitBody: Codable, Sendable {
    public let commit: String
}

public struct PairRevealClientBody: Codable, Sendable {
    public let clientId: String
    public let clientPk: String
    public let nonceC: String
    public let name: String
}

public struct PairRevealHostBody: Codable, Sendable {
    public let hostId: String
    public let hostPk: String
    public let nonceH: String
    public let name: String
}

public struct PairCompleteBody: Codable, Sendable {
    public let hostId: String
    public let clientId: String
}

public enum PairingType {
    public static let ready = "pair.ready"
    public static let commit = "pair.commit"
    public static let revealClient = "pair.reveal_client"
    public static let revealHost = "pair.reveal_host"
    public static let complete = "pair.complete"
}

public enum PairingWire {
    public static func encodeCommit(_ commitment: Data) throws -> String {
        try encode(Envelope(t: PairingType.commit, b: PairCommitBody(commit: commitment.base64EncodedString())))
    }

    public static func encodeRevealHost(identity: LocalIdentity, nonce: Data) throws -> String {
        try encode(
            Envelope(
                t: PairingType.revealHost,
                b: PairRevealHostBody(
                    hostId: identity.deviceId,
                    hostPk: identity.publicKey.base64EncodedString(),
                    nonceH: nonce.base64EncodedString(),
                    name: identity.name
                )
            )
        )
    }

    public static func decodeRevealClient(_ text: String) throws -> PairRevealClientBody {
        try JSONDecoder().decode(Envelope<PairRevealClientBody>.self, from: Data(text.utf8)).b
    }

    public static func decodeComplete(_ text: String) throws -> PairCompleteBody {
        try JSONDecoder().decode(Envelope<PairCompleteBody>.self, from: Data(text.utf8)).b
    }

    public static func type(_ text: String) throws -> String {
        try JSONDecoder().decode(EnvelopeHeader.self, from: Data(text.utf8)).t
    }

    private static func encode<Body: Codable & Sendable>(_ envelope: Envelope<Body>) throws -> String {
        String(data: try JSONEncoder().encode(envelope), encoding: .utf8) ?? ""
    }
}

private struct EnvelopeHeader: Codable {
    let t: String
}

File: android/app/src/main/kotlin/com/edgelink/core/PairingWire.kt
package com.edgelink.core

import kotlinx.serialization.Serializable
import java.util.Base64

@Serializable
data class PairStartRequest(
    val hostId: String,
    val hostPk: String,
    val name: String
)

@Serializable
data class PairClaimRequest(
    val hostId: String,
    val clientId: String,
    val clientPk: String,
    val name: String
)

@Serializable
data class PairConfirmRequest(
    val role: String,
    val hostId: String,
    val clientId: String,
    val hostPk: String,
    val clientPk: String,
    val hostName: String,
    val clientName: String
)

@Serializable
data class PairReadyBody(val deviceId: String)

@Serializable
data class PairCommitBody(val commit: String)

@Serializable
data class PairRevealClientBody(
    val clientId: String,
    val clientPk: String,
    val nonceC: String,
    val name: String
)

@Serializable
data class PairRevealHostBody(
    val hostId: String,
    val hostPk: String,
    val nonceH: String,
    val name: String
)

@Serializable
data class PairCompleteBody(
    val hostId: String,
    val clientId: String
)

object PairingTypes {
    const val READY = "pair.ready"
    const val COMMIT = "pair.commit"
    const val REVEAL_CLIENT = "pair.reveal_client"
    const val REVEAL_HOST = "pair.reveal_host"
    const val COMPLETE = "pair.complete"
}

object PairingWire {
    fun encodeReady(deviceId: String): String =
        EnvelopeCodec.json.encodeToString(Envelope(PairingTypes.READY, PairReadyBody(deviceId)))

    fun encodeCommit(commitment: ByteArray): String =
        EnvelopeCodec.json.encodeToString(
            Envelope(PairingTypes.COMMIT, PairCommitBody(Base64.getEncoder().encodeToString(commitment)))
        )

    fun encodeRevealClient(identity: LocalIdentity, nonce: ByteArray): String =
        EnvelopeCodec.json.encodeToString(
            Envelope(
                PairingTypes.REVEAL_CLIENT,
                PairRevealClientBody(
                    clientId = identity.deviceId,
                    clientPk = Base64.getEncoder().encodeToString(identity.publicKey),
                    nonceC = Base64.getEncoder().encodeToString(nonce),
                    name = identity.name
                )
            )
        )

    fun decodeCommit(text: String): PairCommitBody =
        EnvelopeCodec.json.decodeFromString<Envelope<PairCommitBody>>(text).b

    fun decodeRevealHost(text: String): PairRevealHostBody =
        EnvelopeCodec.json.decodeFromString<Envelope<PairRevealHostBody>>(text).b

    fun decodeComplete(text: String): PairCompleteBody =
        EnvelopeCodec.json.decodeFromString<Envelope<PairCompleteBody>>(text).b

    fun type(text: String): String =
        EnvelopeCodec.json.decodeFromString<Envelope<kotlinx.serialization.json.JsonObject>>(text).t
}
